### ISSUE TITLE
Explicit invocation of action within action_run switch statement

### ISSUE THREAD

**Issue by lzhzero** [p4lang/p4c#2951]
For an action_run switch statement that has codes like:

```
switch (X.apply().action_run) {
  A1 : { Y.apply(); A4(); }
  A2 : { Y.apply(); A4(); }
}
```

If the switch statement contains multiple explicit invocation of a same action
(`A4()` here), it will cause a compiler error: "table Y has multiple successors"

Calling the action using a seperate table will eliminate the compiler error.

```
table Z {
  key = { y : exact;}
  actions = { A4;}
}
switch (X.apply().action_run) {
  A1 : { Y.apply(); Z.apply(); }
  A2 : { Y.apply(); Z.apply(); }
}
```

[test.p4.txt](https://github.com/p4lang/p4c/files/7550204/test.p4.txt)

**Comment by mbudiu-vmw** [comment]
The same explanation given for #2950 applies here as well: on BMV2 you cannot execute actions directly, the compiler will synthesize a table (with no keys) to invoke that action. But the compiler isn't smart enough to understand that the same table could be used for both invocations of action A4, so it will create two different tables. So that's why Y has two different successors.

**Comment by lzhzero** [comment]
The P4_16 spec allows explicit invocation of actions "either from a control block or from another action".
Other people may run into similar issues if they use explicit invocations in some complexed situation similar to the example above.

Could this have an easy fix? I saw that in the front end pass `FrontEnd_50_LocalizeAllActions`, a copy of the action is made, and later in another stage, 2 different tables are created one for each of the action.

**Comment by mbudiu-vmw** [comment]
Yes, the compiler clones actions with the expectation that each clone will be optimized differently depending on how it's used.
I don't remember whether any subsequent passes depend on each action having exactly 1 caller.

Everything can be fixed, but at this point I think that this is a low priority problem compared to other open bugs that we have. If you are interested in contributing code to "fix" this issue we can discuss about how that could be approached.

**Comment by jafingerhut** [comment]
@lzhzero Is it reasonable to close this issue now?  It is true that the BMv2 back end could be generalized, but no one has volunteered to do so, and it seems unlikely given the amount of work involved.
